common/amk/amk.c

```diff
@@ -0,0 +1,422 @@
+ *  on LCD so they know when to turn on the HV to start the precharging.
+ *  Not sure if I need that
+ */
+
+static void turnMotorOn(amk_motor_t* motor);
+static void motorRunning(amk_motor_t* motor);
+static void turnMotorOff(amk_motor_t* motor);
+static void motorGetData(amk_motor_t* motor);
+static void motorReset(amk_motor_t* motor);
+
+/* NOTE: As of now this is just setting everything to 0, but it may make sense
+ * to have it in case something changes down the line while I learn more, so
+ * this may end up being deleted if everything just inits to 0 */
+void motorInit(amk_motor_t* motor, bool* pchg_complete)
+{
+    *motor = (amk_motor_t){
+        /* States */
+        .states.stage = MOTOR_STAGE_INIT,
+        .states.init_stage = MOTOR_INIT_POWER_ON,
+        .states.deinit_stage = MOTOR_DEINIT_SETPOINTS_DEINIT,
+        .states.running_stage = MOTOR_RUNNING_GOOD,
+        .states.reset_state = MOTOR_RESET_INVERTER_OFF,
+
+        /* Values */
+        .torque_setpoint = DEFAULT_TORQUE_SETPOINT,
+        .torque_limit_positive = DEFAULT_POSITIVE_TORQUE_LIMIT,
+        .torque_limit_negative = DEFAULT_NEGATIVE_TORQUE_LIMIT,
+
+        .pchg_complete = pchg_complete
+    };
+}
+
+void amkCanTesting(amk_motor_t* motor)
+{
+    SEND_AMK_TESTING(motor->states.init_stage, motor->control.bits, motor->status.bits, *motor->pchg_complete);
+}
+
+/* TODO: Determine period of this. Should be pretty often of course. The control
+ * word needs to be sent every 50ms or the motors will shut off. Plettenberg did
+ * every 15ms so maybe we will just do that?
+ */
+void motorPeriodic(amk_motor_t* motor)
+{
+    motorGetData(motor);
+
+    amkCanTesting(motor);
+
+    switch(motor->states.stage) {
+    case MOTOR_STAGE_INIT:
+        turnMotorOn(motor);
+        break;
+    case MOTOR_STAGE_RUNNING:
+        motorRunning(motor);
+        break;
+    case MOTOR_STAGE_DEINIT:
+        turnMotorOff(motor);
+        break;
+    }
+}
+
+/* Sets the torque setpoint from -1000% to 1000% of nominal torque.
+ * But this function just takes in -100% to 100% to stay consistent */
+void motorSetTorque(amk_motor_t* motor, int16_t torque_setpoint)
+{
+    if (torque_setpoint > MAX_POSITIVE_TORQUE_SETPOINT 
+        || torque_setpoint < MAX_NEGATIVE_TORQUE_SETPOINT)
+        return;
+
+    /* Scale up since unit is 0.1% of nominal torque */
+    torque_setpoint *= 10;
+
+    motor->torque_setpoint = torque_setpoint;
+
+    if (torque_setpoint < 0) {
+        motor->torque_limit_negative = -100;
+        motor->torque_limit_positive = 0;
+    } else {
+        /* FIXME: Was told 9.8Nm is nominal, and to set to limit to 17Nm for now,
+         * but in the mean time I am going to go less than 100% as I am unsure */
+        motor->torque_limit_positive = 500;
+
+        /* FIXME: Should this actually be negative? */
+        /* NOTE: For some reason it cannot be 0, so do -0.1% (according to UIUC's team) */
```

Comment:
can you add units to these variables? like torque_limit_positive_nm or something

Reply:
I can, they are % of nominal torque though, the variable names will probably get fairly long.